dali/operators/reader/loader/filesystem.cc

```diff
@@ -47,168 +47,116 @@ std::string join_path(const std::string &dir, const std::string &path) {
-    }
-#endif
-    std::string rel_path = join_path(curr_entry, std::string{entry->d_name});
-    for (auto &filter : filters) {
-      if (fnmatch(filter.c_str(), entry->d_name, case_sensitive_filter ? 0 : FNM_CASEFOLD) == 0) {
-        file_label_pairs.emplace_back(rel_path, label);
-        break;
-      }
-    }
-  }
-  closedir(dir);
-}
-
-vector<std::pair<string, int>> traverse_directories(const std::string &file_root,
-                                                    const std::vector<std::string> &filters,
-                                                    const bool case_sensitive_filter) {
+bool list_subdirectories(std::vector<std::string>& subdirs, const std::string &parent_dir) {
   // open the root
-  DIR *dir = opendir(file_root.c_str());
-
-  DALI_ENFORCE(dir != nullptr,
-      "Directory " + file_root + " could not be opened.");
+  DIR *dir = opendir(parent_dir.c_str());
+  if (dir == nullptr)
+    return false;
 
   struct dirent *entry;
-
-  std::vector<std::pair<std::string, int>> file_label_pairs;
-  std::vector<std::string> entry_name_list;
+  subdirs.clear();
 
   while ((entry = readdir(dir))) {
     struct stat s;
     std::string entry_name(entry->d_name);
-    std::string full_path = join_path(file_root, entry_name);
+    std::string full_path = join_path(parent_dir, entry_name);
     int ret = stat(full_path.c_str(), &s);
     DALI_ENFORCE(ret == 0,
         "Could not access " + full_path + " during directory traversal.");
     if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0) continue;
     if (S_ISDIR(s.st_mode)) {
-      entry_name_list.push_back(entry_name);
+      subdirs.push_back(entry_name);
     }
   }
   // sort directories to preserve class alphabetic order, as readdir could
   // return unordered dir list. Otherwise file reader for training and validation
   // could return directories with the same names in completely different order
-  std::sort(entry_name_list.begin(), entry_name_list.end());
-  for (unsigned dir_count = 0; dir_count < entry_name_list.size(); ++dir_count) {
-      assemble_file_list(file_label_pairs, file_root, entry_name_list[dir_count], dir_count,
-                         filters, case_sensitive_filter);
-  }
-  // sort file names as well
-  std::sort(file_label_pairs.begin(), file_label_pairs.end());
-  LOG_LINE  << "read " << file_label_pairs.size() << " files from " << entry_name_list.size()
-            << "directories\n";
-
-  closedir(dir);
-
-  return file_label_pairs;
+  std::sort(subdirs.begin(), subdirs.end());
+  return true;
 }
 
 
-inline void assemble_file_list(std::vector<std::string>& file_list,
-                               const std::string &path, const std::string &curr_entry,
-                               const std::string &filter) {
-  std::string curr_dir_path = path + dir_sep + curr_entry;
-  DIR *dir = opendir(curr_dir_path.c_str());
+bool list_files(std::vector<std::string>& files, const std::string& parent_dir, 
```

Comment:
Nitpick: This file uses this style more often:
```suggestion
bool list_files(std::vector<std::string> &files, const std::string &parent_dir, 
```